sos/upload/targets/__init__.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2024, Jake Hunsaker <jacob.r.hunsaker@gmail.com>
+
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+import getpass
+import logging
+import os
+
+from inspect import getmembers, ismethod
+
+
+class SoSUploadTarget:
+    """
+    This class abstracts destinations for upload into 'targets' that are more
+    human-friendly for reference. An upload target will handle all necessary
+    logic for uploading a given archive/file to the intended destination. The
+    most common of targets will be to vendors that provide end-user support,
+    and in those cases an upload target class should handle any necessary
+    gating as well as authentication or filtering.
+
+    By default, targets will attempt to determine a set of username/password
+    credentials based off the following order:
+
+        1. The setting of `--upload-user` and `--upload-pass` in either the
+           commandline execution of sos or a given config file.
+        2. The setting of the SOSUPLOADUSER and SOSUPLOADPASS env vars
+        3. A prompt for user input, iff the target has set the
+           `prompt_for_credentials` class var to `True`. This prompt will be
+           skipped if `--batch` is set.
```

Comment:
Current ordering is 2., then 1., then 3., due to:

```
    def get_upload_user(self):
        """Helper function to determine if we should use the policy default
        upload user or one provided by the user

        :returns: The username to use for upload
        :rtype: ``str``
        """
        return (os.getenv('SOSUPLOADUSER', None) or
                self.upload_user or
                self._upload_user)
```

method, am I right? Why the change?

(on one side, cmdline option should override more-generic env.variable that is more "persistent", but config file is yet "more persistent" than env.variable; also when user sets both env.var and cmdline option, we should prefer the value that is obfuscated (in process list))